File: anvil/src/main/java/trikita/anvil/RenderableAdapter.java
package trikita.anvil;

import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.FrameLayout;

import java.util.List;

public abstract class RenderableAdapter extends BaseAdapter {
	
	private int currentPosition = -1;

	public interface Item<T> {
		public void view(int index, T item);
	}

	public static <T> RenderableAdapter withItems(final List<T> items, final Item<T> r) {
		return new RenderableAdapter() {
			public int getCount() {
				return items.size();
			}
			public T getItem(int pos) {
				return items.get(pos);
			}
			public void view(int pos) {
				r.view(pos, getItem(pos));
			}
		};
	}

	@Override
	public View getView(int pos, View convertView, ViewGroup parent) {
		View v = convertView;
		if (v == null) {
			currentPosition = pos;
			FrameLayout vg = new FrameLayout(parent.getContext());
			Anvil.Mount m = new Anvil.Mount(vg, new Anvil.Renderable() {
				public void view() {
					RenderableAdapter.this.view(currentPosition);
				}
			});
			m.render();
			vg.setTag(m);
			v = vg;
		} else {
			Anvil.Mount m = (Anvil.Mount) v.getTag();
			currentPosition = pos;
			m.render();
		}
		return v;
	}

	@Override
	public long getItemId(int pos) {
		return pos; // just a most common implementation
	}

	public abstract void view(int index);
}

File: anvil/src/main/kotlin/dev/inkremental/RenderableAdapter.kt
package dev.inkremental

import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.FrameLayout
import dev.inkremental.Inkremental.Mount

abstract class RenderableAdapter : BaseAdapter() {
    private var currentPosition = -1

    override fun getView(pos: Int, convertView: View?, parent: ViewGroup): View {
        var v = convertView
        if (v == null) {
            currentPosition = pos
            val vg = FrameLayout(parent.context)
            val m = Mount(vg) { view(currentPosition) }
            Inkremental.render(m)
            vg.tag = m
            v = vg
        } else {
            val m = v.tag as Mount
            currentPosition = pos
            Inkremental.render(m)
        }
        return v
    }

    override fun getItemId(pos: Int): Long {
        return pos.toLong() // just a most common implementation
    }

    abstract fun view(index: Int)

    companion object {
        fun <T> withItems(items: List<T>, r : (index: Int, item: T) -> Unit): RenderableAdapter {
            return object : RenderableAdapter() {
                override fun getCount(): Int {
                    return items.size
                }

                override fun getItem(pos: Int): T {
                    return items[pos]
                }

                override fun view(pos: Int) {
                    r(pos, getItem(pos))
                }
            }
        }
    }
}